src/btree/bt_sync.c

```diff
@@ -143,6 +143,33 @@ __sync_ref_mark_deleted(WT_SESSION_IMPL *session, WT_REF *ref)
     return (0);
 }
 
+/*
+ * __sync_ref_int_obsolete_cleanup --
+ *     Traverse the internal page and identify the leaf pages that are obsolete and mark them as
+ *     deleted.
+ */
+static int
+__sync_ref_int_obsolete_cleanup(WT_SESSION_IMPL *session, WT_REF *intref)
+{
+    WT_PAGE_INDEX *pindex;
+    WT_REF *ref;
+    uint32_t slot;
+
+    WT_INTL_INDEX_GET(session, intref->page, pindex);
+    for (slot = 0; slot < pindex->entries; slot++) {
+        ref = pindex->index[slot];
+
+        /* Skip the deleted child pages. */
+        if (ref->state == WT_REF_DELETED)
+            continue;
```

Comment:
I think it's odd that we test the case `WT_REF.state == WT_REF_DELETED` here, but other tests like `ref->page->modify->rec_result != WT_PM_REC_EMPTY` appear in `__sync_ref_is_obsolete()`. Why is that? Would it make more sense to put all the tests for "don't mark this page deleted" in one place?

Reply:
I removed this case entirely while processing the internal pages as this case never happens now because we never mark the empty pages as deleted after reconciliation.